lib/using_restricted_array.rb

```diff
@@ -9,53 +11,109 @@
 # Time complexity: ?
 # Space complexity: ?
 def length(array)
-  raise NotImplementedError
+  # array[index]
+  i = 0
+  i += 1 until array[i].nil?
+
+  i
+  # raise NotImplementedError
 end
 
 # Prints each integer values in the array
 # Time complexity: ?
 # Space complexity: ?
 def print_array(array)
-  raise NotImplementedError
+  # raise NotImplementedError
+  i = 0
+  until array[i].nil?
+    print array[i]
+    i += 1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 # Time complexity: ?
 # Space complexity: ?
-def search(array, length, value_to_find)
-  raise NotImplementedError
-end
+def search(array, _length, value_to_find)
+    # raise NotImplementedError
+    i = 0
+    until array[i] == nil
+      if array[i] == value_to_find
+        return true
+      else
+        return false
+      end
+    end
+  end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
 # Time complexity: ?
 # Space complexity: ?
-def find_largest(array, length)
-  raise NotImplementedError
+def find_largest(array, _length)
+  # raise NotImplementedError
+  i = 0
+  largest = array[i]
+  until array[i] == nil
+      if array[i] > largest
+        largest = array[i]
+      end
+    i = i + 1
+  end
+  return largest
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
 # Time complexity: ?
 # Space complexity: ?
-def find_smallest(array, length)
-  raise NotImplementedError
+def find_smallest(array, _length)
+  # raise NotImplementedError
+  i = 0
+  smallest = array[i]
+  until array[i] == nil
+    if array[i] < smallest
+      smallest = array[i]
+    end
+    i = i + 1
+  end
+  return smallest
+
 end
 
 # Reverses the values in the integer array in place
 # Time complexity: ?
 # Space complexity: ?
 def reverse(array, length)
-  raise NotImplementedError
+  i = 0
+  max = length - 1
+  temp = 0
+  until i > max
+    temp = array[i]
+    array[i] = array[max]
+    array[max] = temp
+    i = i + 1
+    max = max - 1
+  end
+  return array
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 # Time complexity: ?
 # Space complexity: ?
 def binary_search(array, length, value_to_find)
-  raise NotImplementedError
+  #raise NotImplementedError
+  i = 0
+  until array[i] == nil
+    if array[i] == value_to_find
+      return true
+    else
+      return false
+    end
+  end
+  #
```

Comment:
Consider starting with this:

```suggestion
  low = 0
  high = length -1
  
  while (low <= high)
    # put your code here
    
   end
```

Reply:
Can you see how to implement binary search from this starting point.  

Step 1:  Find the middle
Step 2:  if the array at the midpoint is higher than what you're looking for, high becomes midpoint - 1
             If the array at the midpoint is lower than what you're looking for low becomes midpoint + 1
            otherwise you found it

If you reach the end of the loop, you didn't find it.